### PR TITLE
Add Row size methods (#3160)

### LINKED ISSUE #3160: Row Format Size Tracking
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

I would like a way to determine the amount of memory used by the row format, so this can be used to inform memory accounting in DataFusion

**Describe the solution you'd like**

I would like the row format to support nested types

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #3160

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow/src/row/interner.rs [hunk 1]
```diff
@@ -157,6 +157,15 @@ impl OrderPreservingInterner {
     pub fn value(&self, key: Interned) -> &[u8] {
         self.values.index(key)
     }
+
+    /// Returns the size of this instance in bytes including self
+    pub fn size(&self) -> usize {
+        std::mem::size_of::<Self>()
+            + self.values.buffer_size()
+            + self.values.buffer_size()
```
Comment (crepererum) [inline arrow/src/row/interner.rs:165]: twice?

**Review by crepererum — APPROVED** [review]

File: arrow/src/row/mod.rs [hunk 2]
```diff
@@ -358,6 +358,14 @@ impl SortField {
     pub fn new_with_options(data_type: DataType, options: SortOptions) -> Self {
         Self { options, data_type }
     }
+
+    /// Return size of this instance in bytes.
+    ///
+    /// Includes the size of `Self`.
+    pub fn size(&self) -> usize {
+        self.data_type.size() + std::mem::size_of::<Self>()
+            - std::mem::size_of::<DataType>()
```
Comment (viirya) [inline arrow/src/row/mod.rs:367]: Why minus size of `DataType`?

Comment (viirya) [reply arrow/src/row/mod.rs:367]: No need to count for `options`?

Comment (tustvold) [reply arrow/src/row/mod.rs:367]: Options is counted by `size_of<Self>` we need to subtract DataType as otherwise it is counted twice

Comment (tustvold) [reply arrow/src/row/mod.rs:367]: It is that, plus any additional memory from nested fields

Comment (viirya) [reply arrow/src/row/mod.rs:367]: Oh, `- std::mem::size_of::<DataType>()` is for `data_type: DataType` which is already counted in by `std::mem::size_of::<Self>()`?

**Review by viirya — APPROVED** [review]

**Pull request merged**